### ISSUE TITLE
Setting warning in configurations to prevent entry of swapped values 

### ISSUE THREAD

**Issue by kamwal** [euroargodev/User-Acceptance-Test-Python-version-of-the-OWC-tool#6]
I have noticed that currently in Matlab and Python version of OWC there is no warning/throw an error to prevent DMQC operator to enter swapped values in small and large: Lon, Lat, Age parts of the Objective Mapping Parameters section. I run Python and Matlab OWC software with swapped values and in both cases, codes are running and producing map_<WNO>.mat file.

Should we introduce in the OWC code some kind of e.g warning/throw an error, to prevent users to accidentally enter swapped values?

**Comment by Teddyzander** [comment]
@apswong @cabanesc Any thoughts on this?

**Comment by cabanesc** [comment]
Yes, I think it is a good idea that swapped values throw an error.

**Comment by gmaze** [comment]
Entering the discussion here, even not being @apswong nor @cabanesc 

This would be great as long as part of a parameters validation scheme.

But how would you detect swapped values ?
If you test for one value being larger than the other ('small' should be smaller than 'large' I guess), then in fact, we could imagine that users would simply enter 2 values, in a list, and the software would automatically assign the larger to the "large" parameter. 
This would also reduce the number of parameter to manage.
For instance ``MAPSCALE_LONGITUDE_LARGE`` and ``MAPSCALE_LONGITUDE_SMALL`` would "merge" toward ``MAPSCALE_LONGITUDE`` with a list of 2 sorted values.

**Comment by edsmall-bodc** [comment]
Okay, so we have two proposed solutions:

1. Throw an error if values are the wrong way round
2. If we detect values are the wrong way round we can swap them (should we tell the user this is the case?)/just use a sorted list

I agree that we can just pass them as sorted lists - it means we would be cutting down on arguments too, which is always nice. 

We can only do this if we NEVER expect users to have to enter a larger scale than their smaller one for any of their parameters. Is this the definitely the case?

**Comment by apswong** [comment]
Hi, I want to check a few things before I reply to this query.  Please give me a few days.  Thanks.

**Comment by edsmall-bodc** [comment]
@apswong No worries, take all the time you need - bug fixes for update_salinity_mapping will not begin until the end of phase 1, and we are flexible on when that time will be

**Comment by apswong** [comment]
Hi, sorry for the late reply.  I would issue a warning if MAPSCALE_*_SMALL > MAPSCALE_*_LARGE; something like "Are you sure? Press Y to proceed, N to escape and correct the entry in the config file."

I would not automatically put the entries into a sorted list and re-assign the entries automatically.  The users should make 100% of the decisions about their parameter entries. For most users, these parameters are rarely changed, so it doesn't take much for them to correct their entries if they happen to make a mistake once every blue moon.  But if it is not a mistake and the users genuinely want to experiment with their parameter entries, they should be free to do so.  All we can do is issue a warning, and if the users confirm Y, that's what they want, then the code should proceed as instructed.

**Comment by edsmall-bodc** [comment]
@apswong Thanks for the detailed response.

Taking that into consideration, the proposed fix therefore would be to place a warning as described above, something along the lines of:

WARNING: <small paramater name> is larger than <large parameter name>. Are you sure you wish to proceed? y/n?

We could even print the parameter values, if that is helpful.

The user is then notified of the oddity, and can either commit to the analysis by entering a "y", "Y", or "yes". Any other entry would cancel the analysis.

Sound okay?

**Comment by apswong** [comment]
Sounds great.  Thanks!

**Comment by gmaze** [comment]
I don't think having the user to answer a warning is a good idea.
Whenever the tests will be run automatically, this will fail or simply makes it much more complicated to run.
A simple warning should be enough to make the user know about this, and assume it.

**Comment by apswong** [comment]
Good point! Yes, I often run the calibration as a batch job overnight with multiple floats - I wouldn't want to be sitting in front of the computer answering Y or N all night long.  So, yes, a simple warning message somewhere in the output will suffice.  Thanks for pointing this out, @gmaze.

**Comment by edsmall-bodc** [comment]
We could set it so the user only has to press yes once, as a compromise? Other wise there is a chance that they could miss this warning since the console already prints all the parameters and each time we move onto a new profile.

**Comment by gmaze** [comment]
This doesn't solve the problem raised above, you still have to press a key ...
We should believe in the user and its ability to read the program output

**Comment by edsmall-bodc** [comment]
Yes, but only once, right at the beginning of the analysis

**Comment by gmaze** [comment]
1 or 10 is the same issue
during an execution in batch mode, like with automatic scripts running over night for instance, this would be an issue, or at least a complication for the user
you still have to manage this std input

**Comment by edsmall-bodc** [comment]
I'll let @apswong, @cabanesc and @kamwal comment on this, since this is a user issue, as you've said.

I think, from a none-DMQC operator perspective, I would rather be warned once about this and have to manually override it than risk missing the warning and running a batch of floats with the wrong analysis parameters set, as

a) It may be hard for an operator to spot incorrect settings post-analysis, and
b) we expect this event to happen so rarely that the warning should be made as obvious as possible, as it is more likely to happen by accident than on purpose.

So as it stands, we have two options:

1. Simply print a warning to the user (which could potentially be missed by the user due to other console outputs).
2. Print the warning once at the very beginning of the analysis, requiring manual override.

I have no strong opinions either way, and it's a very easy change to make, so if we make one decision now and then want to change or minds later, it would not be a big deal.

**Comment by gmaze** [comment]
There is another possibility: let the user defines the level of parameter checks.
A global package option can be set to defined if the user want an error raised or a simple warning if the validation/check of the parameter analysis is raising non blocking issues

**Comment by cabanesc** [comment]
Hi, 
I can't see any obvious reason why I would like to swap small/large scales, other than to experiment with a given float.   In this case, I would be behind the screen, ready to press a key.
If I switched scales by mistake, I would prefer the program to stop rather than  realize later that my analysis was wrong. 
 I would therefore prefer solution 2.  We shouldn't have to manually enter Y/N very often anyway. 

@gmaze: Indeed, it would  be interesting if the level  of the parameter checks could be defined by the user.

**Comment by edsmall-bodc** [comment]
@cabanesc Thanks for the insight.

So a new proposal here, as combination of the two. 

Allow users to change the parameter validation settings, but have the default setting to require manual override (only once) during analysis if we detect some strangeness in the parameter settings (ie. large/small parameters the wrong way round).

Does any other ideas, or doesn't like this idea?

**Comment by apswong** [comment]
Yep, let's go with that.
